src/org/labkey/test/components/ui/grids/QueryGrid.java

```diff
@@ -299,13 +355,181 @@ public QueryGrid clearAllSelections()
+        // will result in the edit alert text and button text being included.
+        // There is a possible issue if the alert text is 'UPDATED'. This text will fade out and may return a -1 when
+        // calling headerText.indexOf(alertText). Trying to minimize that by first getting the alert text then the header text.
+        //
+        // If this proves to be unreliable it might be easier/safer to return the panel header text and let the test do
+        // a .contains() on it to see if it has the expected header.
+        if(elementCache().panelHeader().isDisplayed())
+        {
+            String alertText = getEditAlertText();
+            String headerText = elementCache().panelHeader().getText();
+            headerText = headerText.contains(alertText) ? headerText.substring(alertText.length()) : headerText;
+            String buttonText = "Undo\nSave";
+            headerText = headerText.contains(buttonText) ? headerText.substring(0, headerText.indexOf(buttonText)) : headerText;
+            String appButtonText = "UndoSave";
+            viewName = headerText.contains(appButtonText) ? headerText.substring(0, headerText.indexOf(appButtonText)) : headerText;
+        }
+        else
+        {
+            viewName = "";
+        }
+
+        return viewName.trim();
+    }
+
+    /**
+     * Click the Undo button in the header. Will wait for the grid to update.
+     *
+     * @return This grid.
+     */
+    public QueryGrid clickUndoButton()
+    {
+        WebElement undoButton = Locator.buttonContainingText("Undo").findElement(elementCache().panelHeader());
+
+        // Wait for the grid to update.
+        doAndWaitForUpdate(undoButton::click);
+
+        return this;
+    }
+
+    /**
+     * Is the 'Undo' button visible on the grid.
+     *
+     * @return True if visible, false otherwise.
+     */
+    public boolean isUndoButtonVisible()
+    {
+        return Locator.buttonContainingText("Undo").findWhenNeeded(elementCache().panelHeader()).isDisplayed();
+    }
+
+    /**
+     * Click the Save (View) button in the grid header. If the grid is showing the default view clicking save will show
+     * a {@link SaveGridViewDialog}, otherwise clicking save will save the changes to the view currently applied to the grid.
+     *
+     * @param expectSaveDialog Indicate if a 'Save View' dialog is expected. Should only happen if you are saving the default view.
+     * @return A {@link SaveGridViewDialog} or null if a saved dialog is not expected.
+     */
+    public SaveGridViewDialog clickSaveButton(boolean expectSaveDialog)
+    {
+        WebElement saveButton = Locator.buttonContainingText("Save").findElement(elementCache().panelHeader());
+        saveButton.click();
+
+        if(expectSaveDialog)
+        {
+            return new SaveGridViewDialog(getDriver(), this);
+        }
+        else
+        {
+            return null;
```

Comment:
Is there a way to wait for the save to complete? Maybe wait for the save button to disappear?
```suggestion
            getWrapper().shortWait().until(ExpectedConditions.stalenessOf(saveButton));
            return null;
```

Reply:
Changed.